oemof/outputlib/processing.py

```diff
@@ -0,0 +1,135 @@
+    is fetched as the last element. For time-independent data (scalars)
+    zero ist returned.
+    """
+    if all(issubclass(type(n), Node) for n in x):
+        return 0
+    else:
+        return x[-1]
+
+
+def remove_timestep(x):
+    """
+    Remove the timestep from oemof tuples.
+
+    The timestep is removed from tuples of type `(n, n, int)` and `(n, int)`.
+    """
+    if all(issubclass(type(n), Node) for n in x):
+        return x
+    else:
+        return x[:-1]
+
+
+def create_dataframe(om):
+    """
+    Create a result dataframe with all optimization data.
+
+    Results from Pyomo are written into pandas DataFrame where separate columns
+    are created for the variable index e.g. for tuples of the flows and
+    components or the timesteps.
+    """
+    # get all pyomo variables including their block
+    block_vars = []
+    for bv in om.component_data_objects(Var):
+        block_vars.append(bv.parent_component())
+    block_vars = list(set(block_vars))
+
+    # write them into a dict with tuples as keys
+    var_dict = {(str(bv).split('.')[0], str(bv).split('.')[-1], i): bv[i].value
+                for bv in block_vars for i in getattr(bv, '_index')}
+
+    # use this to create a pandas dataframe
+    df = pd.DataFrame(list(var_dict.items()), columns=['pyomo_tuple', 'value'])
+    df['variable_name'] = df['pyomo_tuple'].str[1]
+
+    # adapt the dataframe by separating tuple data into columns depending
+    # on which dimension the variable/parameter has (scalar/sequence).
+    # columns for the oemof tuple and timestep are created
+    df['oemof_tuple'] = df['pyomo_tuple'].map(get_tuple)
+    df['timestep'] = df['oemof_tuple'].map(get_timestep)
+    df['oemof_tuple'] = df['oemof_tuple'].map(remove_timestep)
+
+    # order the data by oemof tuple and timestep
+    df = df.sort_values(['oemof_tuple', 'timestep'], ascending=[True, True])
+
+    return df
+
+
+def results(es, om):
+    """
+    Create a result dictionary from the result DataFrame.
+
+    Results from Pyomo are written into a dictionary of pandas objects where
+    a Series holds all scalar values and a dataframe all sequences for nodes
+    and flows.
+    The dictionary is keyed by the nodes e.g. `results[(n,)]['scalars']`
+    and flows e.g. `results[(n,n)]['sequences']`.
+    """
+    df = create_dataframe(om)
+
+    # create a dict of dataframes keyed by oemof tuples
+    df_dict = {k: v[['timestep', 'variable_name', 'value']]
+               for k, v in df.groupby('oemof_tuple')}
+
+    # create final result dictionary by splitting up the dataframes in the
+    # dataframe dict into a series for scalar data and dataframe for sequences
+    results = {}
+    for k, v in df_dict.items():
+        df_dict[k].set_index('timestep', inplace=True)
+        df_dict[k] = df_dict[k].pivot(columns='variable_name', values='value')
+        df_dict[k].index = es.timeindex
+        try:
+            condition = df_dict[k].isnull().any()
```

Comment:
in commit 4fefc4bb71f @simnh added a possibility to check the solver output.

Furthermore [chapter 18.10 of the pyomo documentation](https://software.sandia.gov/downloads/pub/pyomo/PyomoOnlineDocs.html#_accessing_solver_status) shows a way to check the status.

```python
from pyomo.opt import SolverStatus, TerminationCondition
if es.results.solver.Solver.Status == SolverStatus.ok) and (es.results.solver.termination_condition == TerminationCondition.optimal):
     logging.error("The solver terminated in an non-optimal state. You cannot process any results)
```

Afterwards you could keep your check but change the message to "Something odd happend..." because you know that it is not caused by the solver because you already checked this case before.